### PR TITLE
Add COSE support to Rekor

### PR DESCRIPTION
**Pull request by kommendorkapten**
#### Summary
Basic support for COSE envelopes.
Added is functionality to upload COSE records, optionally signed with AAD. The AAD is not stored in Rekor.
The COSE envelope is stored as an attestation if attestation storage is enabled, and can so be retrieved via Rekor. The returned envelope is returned in the `attestation` field, not within the `body` field. 

What's stored is the hash of the payload and the entire envelope. If the content type of the COSE envelope is an in-toto statement, the payload is parsed and the subject hashes are extracted and indexed for searching.

#### Ticket Link
This pull requests completes https://github.com/sigstore/rekor/pull/731.

#### Release Note

```release-note
Added new request/response type: COSE envelopes. When uploading a COSE envelope, only the hash of the payload and the hash of the envelope is stored, unless attestation storage is enabled, then the entire envelope is stored as an attestation.
A new command line parameter `--aad` is added. This parameter provides extra Additional Authenticated Data to be used during COSE signature verification (using AAD with COSE is optional).
```

### CODE REVIEW

File: cmd/rekor-cli/app/pflag_groups.go
Comment (bobcallaway) [inline cmd/rekor-cli/app/pflag_groups.go]: you may want to spell out what `aad` means here, since this will be printed as part of command line flag help output.

File: pkg/types/cose/cose.go
Comment (bobcallaway) [inline pkg/types/cose/cose.go]: can you switch to 2022 throughout all new files please?

**Comment by dlorenc** [comment]
Looks like there's a merge conflict, and github isn't running for some reason. Maybe fixing the merge conflict will poke actions?

**Comment by kommendorkapten** [comment]
Yes, will do. Some merge conflicts I have to resolve first 👍

File: pkg/types/cose/v0.0.1/entry.go
Comment (dlorenc) [inline pkg/types/cose/v0.0.1/entry.go]: Do we need to be a bit more careful here to avoid curve mismatches? Is there a possibility someone actually used PS384 or PS512?

File: pkg/types/cose/v0.0.1/entry.go
Comment (dlorenc) [inline pkg/types/cose/v0.0.1/entry.go:108]: This part is interesting, I'm not sure I've seen in-toto statements embedded into COSE envelopes before. I think it makes sense overall though.

**Review by dlorenc — COMMENTED** [review]
One main question on algorithm checking.

File: pkg/types/cose/v0.0.1/entry.go
Comment (kommendorkapten) [reply pkg/types/cose/v0.0.1/entry.go]: Yeah, good catch. I can play around a bit more early next week to see if we can tighten up this.

**Comment by dlorenc** [comment]
Overall LGTM. I'll wait for @bobcallaway to take a look before merging.

**Comment by dlorenc** [comment]
Overall LGTM. I'll wait for @bobcallaway to take a look before merging.

File: go.mod
Comment (bobcallaway) [inline go.mod]: there was a recent effort to stop using this and use `errors` instead.

File: pkg/types/cose/cose.go
Comment (bobcallaway) [inline pkg/types/cose/cose.go]: can you swap this with `"errors"` instead?

File: pkg/types/cose/v0.0.1/entry.go
Comment (bobcallaway) [inline pkg/types/cose/v0.0.1/entry.go]: did you mean to return `err` here?

File: pkg/types/cose/v0.0.1/entry.go
Comment (bobcallaway) [inline pkg/types/cose/v0.0.1/entry.go]: should we ensure unmarshaling is successful before assigning the output to `v.sign1Msg`?

File: pkg/types/entries.go
Comment (bobcallaway) [inline pkg/types/entries.go]: should this get changed to `[]byte` and leave it up to the caller to decode the input correctly?

**Review by bobcallaway — COMMENTED** [review]
looking good overall, just a few questions/nits

File: pkg/types/cose/cose.go
Comment (kommendorkapten) [reply pkg/types/cose/cose.go]: Of course, my bad!

File: pkg/types/cose/v0.0.1/entry.go
Comment (kommendorkapten) [reply pkg/types/cose/v0.0.1/entry.go]: Of course, will go over some test coverage here to make sure we got it covered.

File: pkg/types/cose/v0.0.1/entry.go
Comment (kommendorkapten) [reply pkg/types/cose/v0.0.1/entry.go]: You mean like this:
```go
        coseMsg := gocose.NewSign1Message()
	if err := coseMsg.UnmarshalCBOR(v.CoseObj.Message); err != nil {
		return err
	}
        v.sign1Msg = coseMsg
```

I think that makes sense, we can even wait until the signature is verified until we assign it to `v.sign1Msg`.

File: pkg/types/entries.go
Comment (kommendorkapten) [reply pkg/types/entries.go]: Yes, that makes sense! So basically moving the base64 decoding into the cli part instead of `CreteFromArtifactProperties`. I'll give it a shot to see if I can make it happen.

File: cmd/rekor-cli/app/pflag_groups.go
Comment (bobcallaway) [inline cmd/rekor-cli/app/pflag_groups.go:163]: is there a reason you're squelching this error here?

File: cmd/rekor-cli/app/pflags_test.go
Comment (bobcallaway) [inline cmd/rekor-cli/app/pflags_test.go:350]: maybe one more test case for invalid base64 data passed on the command line?

**Review by bobcallaway — COMMENTED** [review]
LGTM, would like one more test case to ensure we have coverage for non-base64 encoded input on the `aad` cmd line argument

File: pkg/types/cose/v0.0.1/entry.go
Comment (bobcallaway) [reply pkg/types/cose/v0.0.1/entry.go]: yes, ideally I wouldn't want to partially change the pointer receiver unless all validation is successful.

**Review by bobcallaway — DISMISSED** [review]

**Comment by kommendorkapten** [comment]
The sharding e2e test failed (`curl: (22) The requested URL returned error: 422 Unprocessable Entity`). Is there any documentation on running this locally?

**Comment by kommendorkapten** [comment]
Oh, found a script, will try that.

**Comment by kommendorkapten** [comment]
I have run the sharding-e2e test now locally three times and am not able to reproduce, the execution succeeds all the time for me. Anyone having an idea?

**Comment by kommendorkapten** [comment]
Hm, this seems to be the issue:
```
++ printf %x 7051007558037933689
++ Awk '{printf "%016s", $0}'
+ HEX_INITIAL_TREE_ID=61da360425720e79
++ printf %x 151938258151519870
++ awk '{printf "%016s", $0}'
+ HEX_INITIAL_SHARD_ID=' 21bcb0a70e9b67e'
+ ENTRY_ID_1=61da360425720e79d2f305428d7c222d7b77f56453dd4b6e6851752ecacc78e5992779c8f9b61dd9
+ ENTRY_ID_2=' 21bcb0a70e9b67e59a575f157274702c38de3ab1e1784226f391fb79500ebf9f02b4439fb77574c'
++ curl -f http://localhost:3000/api/v1/log/entries/retrieve -H 'Content-Type: application/json' -H 'Accept: application/json' -d '{ "entryUUIDs": ["61da360425720e79d2f305428d7c222d7b77f56453dd4b6e6851752ecacc78e5992779c8f9b61dd9", " 21bcb0a70e9b67e59a575f157274702c38de3ab1e1784226f391fb79500ebf9f02b4439fb77574c"]}'
++ jq '. | length'
```

The shard id gets a white space (` `) as the first character, which is then input to the UUIDs: `" 21bcb0a70e9b67e59a575f157274702c38de3ab1e1784226f391fb79500ebf9f02b4439fb77574c"`.

Running the printf/awk on my machine (macOS) I get a different output
```
 printf %x 151938258151519870 | awk '{printf "%016s", $0}'
021bcb0a70e9b67e
```
Running same command on Ubuntu:
```
# printf %x 151938258151519870 | awk '{printf "%016s", $0}'
 21bcb0a70e9b67e
```
 I'll prepare a patch for the awk script.

**Comment by bobcallaway** [comment]
I'll push up a separate PR to fix that, and also add log collection to
sharding-e2e-test.sh since if there is a failure right now we don't capture
the docker-compose logs.

**Comment by kommendorkapten** [comment]
From the GNU awk manual (https://www.gnu.org/software/gawk/manual/html_node/Format-Modifiers.html): (emphasis mine)
> A leading ‘0’ (zero) acts as a flag indicating that output should be padded with zeros instead of spaces. This applies only to the *numeric output formats*. This flag only has an effect when the field width is wider than the value to print.

**Comment by kommendorkapten** [comment]
Shamelessly copied this from stackoverflow (some minor modification to work with `printf` instead of `echo`):
```
printf abc123 | awk '{ for(c = 0; c < 16 ; c++) s = s"0"; s = s$1; print substr(s, 1 + length(s) - 16);}'
0000000000abc123
```

Tested on both Ubuntu and macOS

**Comment by bobcallaway** [comment]
> Shamelessly copied this from stackoverflow (some minor modification to work with `printf` instead of `echo`):
> 
> ```
> printf abc123 | awk '{ for(c = 0; c < 16 ; c++) s = s"0"; s = s$1; print substr(s, 1 + length(s) - 16);}'
> 0000000000abc123
> ```
> 
> Tested on both Ubuntu and macOS

@priyawadhwa @lkatalin is the intent for the value to be left-padded with zeros?

**Comment by lkatalin** [comment]
> @priyawadhwa @lkatalin is the intent for the value to be left-padded with zeros?

@bobcallaway 
Still going through trying to understand this error, but the [padding](https://github.com/sigstore/rekor/blob/45da1afab938270c881d0ce2cee8b2169fb51289/pkg/sharding/sharding.go#L85) is something I had included so that we can get a [consistent length](https://github.com/sigstore/rekor/blob/45da1afab938270c881d0ce2cee8b2169fb51289/pkg/sharding/sharding.go#L32), probably so that we can easily parse the TreeID from the UUID.

**Review by dlorenc — APPROVED** [review]

**Pull request merged**
